sql: TestQueryProgress failed

[(sql).TestQueryProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2154669&tab=buildLog) on [release-20.1@22875686ba62a08873b057edbcbcf4287f9bb826](https://github.com/cockroachdb/cockroach/commits/22875686ba62a08873b057edbcbcf4287f9bb826):

```
I200806 05:21:36.522137 157129 server/config.go:581  [n?] RocksDB cache size: 128 MiB
I200806 05:21:36.522147 157129 server/config.go:581  [n?] store 0: in-memory, size 0 B
I200806 05:21:36.522289 157129 server/server.go:1514  [n?] no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.
I200806 05:21:36.522328 157129 server/server.go:1524  [n?] **** add additional nodes by specifying --join=127.0.0.1:38151
I200806 05:21:36.527862 157129 server/node.go:324  [n?] **** cluster 171f32cd-56f4-4279-b69c-cec2b3bbb1c8 has been created
I200806 05:21:36.528082 157129 gossip/gossip.go:395  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:38151" > attrs:<> locality:<> ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:0 > build_tag:"v20.1.4-29-g2287568" started_at:1596691296527992985 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:33945" > 
W200806 05:21:36.542440 157254 kv/kvserver/store.go:1631  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200806 05:21:36.542980 157129 server/node.go:428  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=16 KiB), ranges=32, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=166.00 pMax=13304.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I200806 05:21:36.543095 157129 kv/kvserver/stores.go:247  [n1] read 0 node addresses from persistent storage
I200806 05:21:36.543549 157129 server/node.go:649  [n1] connecting to gossip network to verify cluster ID...
I200806 05:21:36.543575 157129 server/node.go:669  [n1] node connected via gossip and verified as part of cluster "171f32cd-56f4-4279-b69c-cec2b3bbb1c8"
I200806 05:21:36.543631 157129 server/node.go:512  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200806 05:21:36.543751 157129 server/server.go:1630  [n1] starting https server at 127.0.0.1:42517 (use: 127.0.0.1:42517)
I200806 05:21:36.543787 157129 server/server.go:1635  [n1] starting postgres server at 127.0.0.1:33945 (use: 127.0.0.1:33945)
I200806 05:21:36.543813 157129 server/server.go:1637  [n1] starting grpc server at 127.0.0.1:38151
I200806 05:21:36.543849 157129 server/server.go:1638  [n1] advertising CockroachDB node at 127.0.0.1:38151
I200806 05:21:36.550718 157277 sql/temporary_schema.go:452  [n1] skipping temporary object cleanup run as it is not the leaseholder
I200806 05:21:36.550745 157277 sql/temporary_schema.go:453  [n1] completed temporary object cleanup job
I200806 05:21:36.550767 157277 sql/temporary_schema.go:575  [n1] temporary object cleaner next scheduled to run at 2020-08-06 05:51:36.545152203 +0000 UTC
W200806 05:21:36.594095 157254 kv/kvserver/store.go:1631  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200806 05:21:36.616886 157331 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200806 05:21:36.632647 157411 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1 User:root}
I200806 05:21:36.655452 157764 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:2a9bd128-06fd-4765-acb2-2c6ef4397930 User:root}
I200806 05:21:36.663678 157844 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200806 05:21:36.668643 157774 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200806 05:21:36.705951 157129 server/server.go:1798  [n1] done ensuring all necessary migrations have run
I200806 05:21:36.705988 157129 server/server.go:2040  [n1] serving sql connections
I200806 05:21:36.706256 157611 sqlmigrations/migrations.go:668  [n1] starting wait for upgrade finalization before schema change job migration
I200806 05:21:36.707087 157611 sqlmigrations/migrations.go:700  [n1] starting schema change job migration
I200806 05:21:36.711592 157612 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:38151 Attrs: Locality: ServerVersion:20.1 BuildTag:v20.1.4-29-g2287568 StartedAt:1596691296527992985 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:33945} ClusterID:171f32cd-56f4-4279-b69c-cec2b3bbb1c8 StartedAt:1596691296527992985 LastUp:1596691296527992985}
I200806 05:21:36.712626 157609 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200806 05:21:36.715709 157611 sqlmigrations/migrations.go:1124  [schema-change-job-migration] evaluating tables for creating jobs
I200806 05:21:36.715780 157611 sqlmigrations/migrations.go:719  [n1] schema change job migration completed
I200806 05:21:36.723692 158008 sql/event_log.go:132  [n1,client=127.0.0.1:53816,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.stats.automatic_collection.enabled Value:false User:root}
I200806 05:21:36.728541 158008 sql/event_log.go:132  [n1,client=127.0.0.1:53816,hostssl,user=root] Event: "create_database", target: 52, info: {DatabaseName:t Statement:CREATE DATABASE t User:root}
I200806 05:21:36.738585 158008 sql/event_log.go:132  [n1,client=127.0.0.1:53816,hostssl,user=root] Event: "create_table", target: 53, info: {TableName:t.public.test Statement:CREATE TABLE t.public.test (x INT8 PRIMARY KEY) User:root}
I200806 05:21:36.741139 158126 kv/kvserver/replica_command.go:397  [n1,split,s1,r32/1:/{Table/36-Max}] initiating a split of this range at key /Table/53 [r33] (zone config)
I200806 05:21:36.780682 158296 jobs/registry.go:806  [n1] job 578751354536361985: stepping through state running with error <nil>
I200806 05:21:36.807282 158296 jobs/registry.go:806  [n1] job 578751354536361985: stepping through state succeeded with error <nil>
I200806 05:21:36.813314 158296 jobs/registry.go:961  [n1] job 578751354536361985: status succeeded after adoption finished
I200806 05:21:36.854258 157129 util/stop/stopper.go:539  quiescing
W200806 05:21:36.854649 158286 kv/kvserver/intentresolver/intent_resolver.go:745  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Table/36/RangeDescriptor: node unavailable; try another peer
--- FAIL: TestQueryProgress (0.41s)
    conn_executor_test.go:544: waiting for query to make progress...
    conn_executor_test.go:504: stalling on scan 6 at /Table/53/1/751/0 and waiting for test to unblock...
    conn_executor_test.go:546: query is now stalled. checking progress...
    conn_executor_test.go:562: 
        	Error Trace:	conn_executor_test.go:562
        	Error:      	Expect "executing" to match "executing \([45]\d\."
        	Test:       	TestQueryProgress

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestQueryProgress PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #51356 sql: TestQueryProgress failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/conn_executor_test.go
+++ b/pkg/sql/conn_executor_test.go
@@ -32,6 +32,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/row"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
@@ -522,7 +523,13 @@ func TestQueryProgress(t *testing.T) {
 	db.Exec(t, `CREATE STATISTICS __auto__ FROM t.test`)
 	const query = `SELECT count(*) FROM t.test WHERE x > $1 and x % 2 = 0`
 
-	ctx, cancel := context.WithCancel(context.Background())
+	// Invalidate the stats cache so that we can be sure to get the latest stats.
+	var tableID sqlbase.ID
+	ctx := context.Background()
+	require.NoError(t, rawDB.QueryRow(`SELECT id FROM system.namespace WHERE name = 'test'`).Scan(&tableID))
+	s.ExecutorConfig().(sql.ExecutorConfig).TableStatsCache.InvalidateTableStats(ctx, tableID)
+
+	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	g := ctxgroup.WithContext(ctx)
@@ -558,8 +565,8 @@ func TestQueryProgress(t *testing.T) {
 	// Although we know we've scanned ~50% of what we'll scan, exactly when the
 	// meta makes its way back to the receiver vs when the progress is checked is
 	// non-deterministic so we could see 47% done or 53% done, etc. To avoid being
-	// flaky, we just make sure we see one of 4x% or 5x%
-	require.Regexp(t, `executing \([45]\d\.`, progress)
+	// flaky, we just make sure we see one of 3x%, 4x% or 5x%
+	require.Regexp(t, `executing \([345]\d\.`, progress)
 }
 
 // This test ensures that when in an explicit transaction, statement preparation

